File: algorithm/JavaScript/ArithmeticSlicesIISubsequence.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var numberOfArithmeticSlices = function(nums) {
    const len = nums.length;
    const f = new Map();
    for (let i = 0; i < len; i++) {
      f[i] = new Map();
    }
    let res = 0;
    for (let i = 0; i < len; i++) {
      for (let j = 0; j < i; j++) {
        const gap = nums[i] - nums[j];
        cnt = f[j].get(gap) || 0;
        res += cnt;
        f[i].set(gap, ( f[i].get(gap) || 0 ) + cnt + 1);
      }
    }

    return res

};

File: algorithm/python/ArithmeticSlicesIISubsequence.py
from collections import defaultdict
class Solution:
    def numberOfArithmeticSlices(self, nums):
        n = len(nums)
        f = [defaultdict(int) for _ in nums]
        res = 0
        for i in range(n):
            for j in range(i):
                gap = nums[i] - nums[j]
                cnt = f[j][gap]
                res += cnt
                f[i][gap] += cnt + 1
        
        return res